auth/authenticator.go

```diff
@@ -0,0 +1,225 @@
+	URL, Username, Password string
+	Client                  *http.Client
+	session                 *session
+}
+
+// GetAuthenticatorFromEnvironment instantiates an Authenticator
+// using service properties retrieved from external config sources.
+func GetAuthenticatorFromEnvironment(credentialKey string) (authenticator core.Authenticator, err error) {
+	props, err := core.GetServiceProperties(credentialKey)
+	if err != nil {
+		return
+	}
+	if props[core.PROPNAME_AUTH_TYPE] == "COUCHDB_SESSION" {
+		authenticator, err = NewCouchDbSessionAuthenticatorFromMap(props)
+		return
+	}
+
+	return core.GetAuthenticatorFromEnvironment(credentialKey)
+}
+
+// NewCouchDbSessionAuthenticator constructs a new NewCouchDbSessionAuthenticator instance.
+func NewCouchDbSessionAuthenticator(url, username, password string) (*CouchDbSessionAuthenticator, error) {
+	authenticator := &CouchDbSessionAuthenticator{
+		URL:      url,
+		Username: username,
+		Password: password,
+	}
+	if err := authenticator.Validate(); err != nil {
+		return nil, err
+	}
+	return authenticator, nil
+}
+
+func NewCouchDbSessionAuthenticatorFromMap(props map[string]string) (*CouchDbSessionAuthenticator, error) {
+	if props == nil {
+		return nil, fmt.Errorf(core.ERRORMSG_PROPS_MAP_NIL)
+	}
+	url := props[core.PROPNAME_SVC_URL]
+	username := props[core.PROPNAME_USERNAME]
+	password := props[core.PROPNAME_PASSWORD]
+	return NewCouchDbSessionAuthenticator(url, username, password)
+}
+
+// AuthenticationType returns the authentication type for this authenticator.
+func (a CouchDbSessionAuthenticator) AuthenticationType() string {
+	return AUTHTYPE_COUCHDB_SESSION
+}
+
+// Validate the authenticator's configuration.
+// Ensures the service url, username and password are valid and not nil.
+func (a CouchDbSessionAuthenticator) Validate() error {
+	// this is what an empty URL ends up after NewCouchDbSessionAuthenticator
+	if a.URL == "" {
+		return fmt.Errorf(core.ERRORMSG_PROP_MISSING, "URL")
+	}
+
+	if !strings.HasPrefix(a.URL, "http") {
+		return fmt.Errorf(core.ERRORMSG_PROP_INVALID, "URL")
+	}
+
+	if a.Username == "" {
+		return fmt.Errorf(core.ERRORMSG_PROP_MISSING, "Username")
+	}
+
+	if a.Password == "" {
+		return fmt.Errorf(core.ERRORMSG_PROP_MISSING, "Password")
+	}
+
+	if core.HasBadFirstOrLastChar(a.Username) {
+		return fmt.Errorf(core.ERRORMSG_PROP_INVALID, "Username")
+	}
+
+	if core.HasBadFirstOrLastChar(a.Password) {
+		return fmt.Errorf(core.ERRORMSG_PROP_INVALID, "Password")
+	}
+
+	return nil
+}
+
+// Authenticate adds session authentication cookie to a request.
+func (a *CouchDbSessionAuthenticator) Authenticate(request *http.Request) error {
+	cookie, err := a.getCookie()
+	if err != nil {
+		return err
+	}
+
+	request.AddCookie(cookie)
+	return nil
+}
+
+// getCookie returns an AuthSession cookie to be used in a request.
+// Whenever a new cookie is needed it fetched from the session end-point.
+func (a *CouchDbSessionAuthenticator) getCookie() (*http.Cookie, error) {
```

Comment:
There is no requirement _per se_. It's nice to follow the same patterns because it makes it easier for us to recollect across all the langs. It's also nice to use core code where possible because:
1. It gives us less of a maintenance problem (if, say, a fix is made in the core we're less likely to notice/consume that fix if we have duplicated).
1. The code gets more coverage/exposure when it is used by more SDKs & authenticator types so bugs are more likely to be found/flushed from it.

That said there is a balance to be struck between the complexity of trying to reuse core code vs the complexity of the code itself, which in reality differs for every language.